[Core] Support Multiple Resources

MaoZiming opened the pull request:
<!-- Describe the changes in this PR -->

This partially builds upon an old PR #1389. 
The user can specify an ordered _preference list_ or an unordered _restriction set_ for accelerators.
_Ordered preference list_. The user prefers `A100-40GB:1` over `V100:1`, etc. The optimizer will follow user-specified order in picking the accelerators.
```
accelerators: ['A100-40GB:1', 'V100:1', 'K80:1', 'T4:1']
```
_Unordered restriction set_. The user specifies multiple accelerators for accelerator to choose from, with no preference among them.
```
accelerators: {'A100-40GB:1', 'K80:1', 'V100:1', 'T4:1', 'T4:4'}
```
<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] New manual tests: `sky launch examples/multi_resources.yaml`
- [x] New manual tests: `sky launch examples/multi_accelerators.yaml` 
- [x] New smoke test: `pytest tests/test_smoke.py::test_multiple_resources_unordered`
- [x] New smoke test: `pytest tests/test_smoke.py::test_multiple_resources_ordered`
- [ ] All smoke tests: `pytest tests/test_smoke.py`

MaoZiming commented:
@infwinston We need to think about the interface.
The optimizer will only print _one_ `chosen_resource` per cloud. If the user specifies multiple accelerators, only one of them (picked by the optimizer) will be printed.

concretevitamin commented:
This is awesome @MaoZiming. Tried it out real quick, some UX notes:

```console
» sky launch --use-spot --down t-multi.yaml
Task from YAML spec: t-multi.yaml
Task(run=<empty>)
  resources: {<Cloud>([Spot], {'A100-80GB': 1}), <Cloud>([Spot], {'A100-40GB': 1})}
I 08-31 10:42:49 optimizer.py:129] Using user-specified resource list.
I 08-31 10:42:49 optimizer.py:1135] No resource satisfying <Cloud>([Spot], {'A100-40GB': 1}) on [AWS, Azure, GCP, Lambda, OCI, SCP].
I 08-31 10:42:49 optimizer.py:1135] No resource satisfying <Cloud>([Spot], {'A100-40GB': 1}) on [AWS, Azure, GCP, Lambda, OCI, SCP].
I 08-31 10:42:49 optimizer.py:703] == Optimizer ==
I 08-31 10:42:49 optimizer.py:715] Target: minimizing cost
I 08-31 10:42:49 optimizer.py:726] Estimated cost: $1.9 / hour
I 08-31 10:42:49 optimizer.py:726]
I 08-31 10:42:49 optimizer.py:800] Considered resources (1 node):
I 08-31 10:42:49 optimizer.py:855] ------------------------------------------------------------------------------------------------------
I 08-31 10:42:49 optimizer.py:855]  CLOUD   INSTANCE               vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE      COST ($)   CHOSEN
I 08-31 10:42:49 optimizer.py:855] ------------------------------------------------------------------------------------------------------
I 08-31 10:42:49 optimizer.py:855]  GCP     a2-ultragpu-1g[Spot]   12      170       A100-80GB:1    europe-west4-a   1.86          ✔
I 08-31 10:42:49 optimizer.py:855] ------------------------------------------------------------------------------------------------------
I 08-31 10:42:49 optimizer.py:855]
Launching a new cluster 'sky-957c-zongheng'. Proceed? [Y/n]:
```

I intentionally used an invalid GPU name and it worked out well by skipping it!
```yaml
resources:
  accelerators: ['A100-40GB:1', 'A100-80GB']
```

1. This set-based repr is surprising, because previously we used a set to mean "pick the best of these candidates" (see `example_app.py`), rather than "try these sequentially":
```
Task(run=<empty>)
  resources: {<Cloud>([Spot], {'A100-80GB': 1}), <Cloud>([Spot], {'A100-40GB': 1})}
```

2. We may want to change this 
```
I 08-31 10:42:49 optimizer.py:129] Using user-specified resource list. 
```
to bold/colored,
```
I 08-31 10:42:49 optimizer.py:129] Using user-specified accelerators list (will be tried in the listed order).
```

3. A bit surprising to see `No resource satisfying <Cloud>([Spot], {'A100-40GB': 1}) on [AWS, Azure, GCP, Lambda, OCI, SCP].` being printed twice?

MaoZiming commented:
@concretevitamin  Thanks for the suggestions!
I have fixed 2 and 3. For 1, we are still deciding on the interface. We initially wanted to use 
`['A100-40GB:1', 'A100-80GB']` to mean a _list_ of accelerators that will be considered sequentially.

MaoZiming commented:
@infwinston hey this should be working with the spot controller. When you get some time maybe could you review the code? Thanks!

concretevitamin commented:
@MaoZiming Quick question: Does the underlying Python API in this PR support https://github.com/skypilot-org/skypilot/issues/2051 https://github.com/skypilot-org/skypilot/issues/1912? Note that in those requests, it's not entirely confirmed users want a preference ordered list. It seems more like a set to support restricting search space.

MaoZiming commented:
@concretevitamin yes. The pr support both an ordered preference list and an unordered restriction set.
I haven't finished the support for region yet, only accelerators. It is tricky to support the list (ordered/unordered) for both _region_ and _accelerators_. One way is to require user to specify either region or accelerators.

infwinston commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -2640,38 +2639,78 @@ def check_resources_fit_cluster(self, handle: CloudVmRayResourceHandle,
+            valid_resource = None
+            requested_resource_list = []
+            for resource in task.resources:
+                task_resources = resource
+                if (task.num_nodes <= handle.launched_nodes and
+                        task_resources.less_demanding_than(
+                            launched_resources,
+                            requested_num_nodes=task.num_nodes)):
```
why do we need a new variable `task_resources` here?

infwinston commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -2640,38 +2639,78 @@ def check_resources_fit_cluster(self, handle: CloudVmRayResourceHandle,
+
+            valid_resource = None
+            requested_resource_list = []
+            for resource in task.resources:
```
I'm not super sure whether the validation is safe. cc @Michaelvll

MaoZiming replied:
Updated

infwinston commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -4194,13 +4233,15 @@ def _check_existing_cluster(
+                task_cloud = r.cloud
+            else:
+                task_cloud = clouds.Cloud
+            if task_cloud is not None:
```
Would `task_cloud` be None?

MaoZiming replied:
removed `if task_cloud is not None:`

infwinston commented on sky/execution.py:
```diff
@@ -628,18 +628,22 @@ def spot_launch(
+        for res in task_.get_resources_list():
+
+            override_params: Dict[str, Any] = {}
+            if not task_.is_resources_mixed_spot:
```
should we raise error if resources are mixed with spot & on-demand?

MaoZiming replied:
I will remove this part and defer it to a later PR>

MaoZiming commented on examples/multi_resources.yaml:
```diff
@@ -0,0 +1,13 @@
+  spot_recovery: FAILOVER
+  # accelerators: ['A100-40GB:1', 'A100-80GB']
+  accelerators: ['A100-40GB:1', 'T4:1', 'V100:1', 'K80:1']
+  # accelerators: K80:1
```
yeah it is intentional to give an example for ordered list and unordered list

MaoZiming commented on examples/multi_resources.py:
```diff
@@ -0,0 +1,71 @@
+                      zone='us-east-1d',
+                      use_spot=True,
+                      accelerators={'K80': 1}),
+        sky.Resources(sky.GCP(), accelerators={'K80': 1}),
```
okay!

MaoZiming commented on sky/backends/backend_utils.py:
```diff
@@ -1927,9 +1927,10 @@ def check_can_clone_disk_and_override_task(
-    task_resources = list(task.resources)[0]
-    if handle.launched_resources.disk_size > task_resources.disk_size:
+    task_resources = task.get_resources_list()
+    first_task_resource = task_resources[0]
```
nice catch. Thanks!

MaoZiming commented on examples/multi_resources.yaml:
```diff
@@ -0,0 +1,22 @@
+
+run: |
+  nvidia-smi
+  sleep 600
```
removed

MaoZiming commented on sky/task.py:
```diff
@@ -556,12 +607,40 @@ def set_resources(
         Returns:
           self: The current task, with resources set.
         """
+        # Reset the preference list.
```
Thanks. Updated

MaoZiming commented on examples/multi_resources.yaml:
```diff
@@ -0,0 +1,22 @@
+  # zone: us-east-1d
+  use_spot: True
+
+  # spot_recovery: FAILOVER
+  spot_recovery: EAGER_NEXT_REGION
```
We don't need to if we are using `sky spot launch...`
I was testing with `python -u -m sky.spot.controller examples/multi_resources.yaml` and I think for that the default `spot_recovery` won't be set.

MaoZiming replied:
Will add the default spot recovery in `StrategyExecutor.make`

infwinston replied:
we expect user to sky launch this yaml so let's make it runnable?

MaoZiming commented on examples/multi_resources.py:
```diff
@@ -0,0 +1,69 @@
+
+task = sky.Task(run='nvidia-smi')
+
+if False:
```
Updated!

MaoZiming commented on sky/optimizer.py:
```diff
@@ -313,6 +354,10 @@ def _estimate_nodes_cost_or_time(
+                elif num_resources != 1:
+                    logger.debug(
+                        'Defaulting the task\'s estimated time to 1 hour.')
+                    estimated_runtime = 1 * 3600
```
just run. the results are the same (this branch and the master branch)

MaoZiming commented on examples/multi_resources.py:
```diff
@@ -0,0 +1,71 @@
+import sky
```
Added. Thanks for the suggestions!

concretevitamin commented on sky/optimizer.py:
```diff
@@ -113,11 +123,25 @@ def optimize(dag: 'dag_lib.Dag',
+                    f'{colorama.Fore.YELLOW}Using user-specified accelerators list{colorama.Style.RESET_ALL} (will be tried in the listed order).'  # pylint: disable=line-too-long
+                )
+                resources_list = dag.tasks[0].get_resources_list()
+                logger.info([
```
Nit: may want to print this with L129. This will make the output log use 1 line vs. 2.

MaoZiming replied:
Thanks updated.

infwinston commented on examples/multi_resources.yaml:
```diff
@@ -0,0 +1,26 @@
+
+  # Ordered Accelerators List
+  resources_ordered: True
+  accelerators: ['A100-40GB:1', 'V100:1', 'T4:1', 'K80:1']
```
We probably need to check the semantic of `sky exec` when multiple resources are specified.

infwinston commented on sky/execution.py:
```diff
@@ -632,6 +632,19 @@ def spot_launch(
     dag_utils.fill_default_spot_config_in_dag(dag)
 
     for task_ in dag.tasks:
+
```
double check if we need to overwrite the params here.

MaoZiming replied:
edited

infwinston commented on sky/optimizer.py:
```diff
@@ -114,11 +124,25 @@ def optimize(dag: 'dag_lib.Dag',
-                minimize_cost=minimize == OptimizeTarget.COST,
-                blocked_resources=blocked_resources,
-                quiet=quiet)
+
```
assert `len(dag.tasks) == 1`?

infwinston commented on sky/optimizer.py:
```diff
@@ -256,15 +280,30 @@ def _estimate_nodes_cost_or_time(
                 # Convert partial resource labels to launchable resources.
-                launchable_resources, cloud_candidates, fuzzy_candidates = (
-                    _fill_in_launchable_resources(node, blocked_resources))
+                launchable_resources, cloud_candidates, fuzzy_candidates = \
```
Double check if master runs the below code elsewhere?

infwinston commented on sky/optimizer.py:
```diff
@@ -784,7 +823,8 @@ def format_number(x):
                                             float]] = {}
             for resources, cost in v.items():
-                cloud = str(resources.cloud)
+                cloud = str(resources.cloud) + resources.get_accelerators_str(
```
is `resource_table_key` a better name?

infwinston commented on sky/optimizer.py:
```diff
@@ -813,7 +861,15 @@ def format_number(x):
+                rows = sorted(
+                    rows,
+                    key=lambda x: (
+                        resources_pref_list.index(x[-4] + (  # pylint: disable=cell-var-from-loop
+                            '[Spot]' if '[Spot]' in x[1] else '')),  # pylint: disable=cell-var-from-loop
+                        x[-2]))  # pylint: disable=cell-var-from-loop
```
add comments on what's `x[-4]` and `x[-2]`

infwinston commented on sky/optimizer.py:
```diff
@@ -919,7 +1045,8 @@ def _cloud_in_list(cloud: clouds.Cloud, lst: Iterable[clouds.Cloud]) -> bool:
 def _make_launchables_for_valid_region_zones(
-    launchable_resources: resources_lib.Resources
+    launchable_resources: resources_lib.Resources,
+    required_regions: Optional[List[str]] = None,
```
is this better?
```suggestion
    allowed_regions: Optional[List[str]] = None,
```

infwinston commented on sky/optimizer.py:
```diff
@@ -1035,13 +1170,18 @@ def _fill_in_launchable_resources(
             if len(launchable[resources]) == 0:
                 clouds_str = str(clouds_list) if len(clouds_list) > 1 else str(
                     clouds_list[0])
+
+                if not print_logger:
+                    continue
+
```
let's double check why we need this?

infwinston commented on sky/resources.py:
```diff
@@ -1009,7 +1034,9 @@ def copy(self, **override) -> 'Resources':
             spot_recovery=override.pop('spot_recovery', self.spot_recovery),
             disk_size=override.pop('disk_size', self.disk_size),
-            region=override.pop('region', self.region),
+            region=override.pop(
+                'region',
+                self.region if self.region_list is None else self.region_list),
```
should we try to refactor `region` as List?

infwinston commented on sky/spot/controller.py:
```diff
@@ -165,6 +164,7 @@ def _run_one_task(self, task_id: int, task: 'sky.Task') -> bool:
             # uninitialized columns.
             spot_state.set_started(job_id=self._job_id,
                                    task_id=task_id,
+                                   resources_str='',
```
why do we need this?

MaoZiming replied:
removed

infwinston commented on sky/spot/controller.py:
```diff
@@ -201,9 +201,18 @@ def _run_one_task(self, task_id: int, task: 'sky.Task') -> bool:
         remote_job_submitted_at = self._strategy_executor.launch()
         assert remote_job_submitted_at is not None, remote_job_submitted_at
 
+        (_, handle) = backend_utils.refresh_cluster_status_handle(
```
Why do we need to specify resource_str manually?

infwinston commented on sky/spot/recovery_strategy.py:
```diff
@@ -411,17 +417,33 @@ def recover(self) -> float:
+                    else:
+                        new_resources_list.append(r.copy())
+
+                task.set_resources(new_resources_list,
```
`[chosen_resource]`?

infwinston commented on sky/spot/recovery_strategy.py:
```diff
@@ -500,6 +522,8 @@ def recover(self) -> float:
         if self._launched_cloud_region is not None:
             task = self.dag.tasks[0]
             requested_resources = list(task.resources)[0]
+            if task.best_resources is not None:
+                requested_resources = task.best_resources
```
why do we need this?

infwinston commented on sky/spot/spot_state.py:
```diff
@@ -317,16 +317,18 @@ def set_starting(job_id: int, task_id: int, callback_func: CallbackType):
 
 def set_started(job_id: int, task_id: int, start_time: float,
-                callback_func: CallbackType):
+                resources_str: str, callback_func: CallbackType):
```
do we need this?

MaoZiming replied:
removed

infwinston commented on sky/task.py:
```diff
@@ -422,10 +423,38 @@ def from_yaml_config(
+                accelerators = {f'{k}:{v}' for k, v in accelerators.items()}
+
+            tmp_resources_list = []
+            for acc in accelerators:
```
add comments on why we need this for loop?

infwinston commented on sky/task.py:
```diff
@@ -422,10 +423,38 @@ def from_yaml_config(
+                tmp_resources_list.append(
+                    sky.Resources.from_yaml_config(tmp_resource))
+
+            if isinstance(accelerators, set):
+                task.set_resources(tmp_resources_list, resources_ordered)
+            elif isinstance(accelerators, list):
+                task.set_resources(tmp_resources_list, resources_ordered)
```
merge this?

infwinston commented on sky/task.py:
```diff
@@ -422,10 +423,38 @@ def from_yaml_config(
+            elif isinstance(accelerators, list):
+                pass
+            elif isinstance(accelerators, dict):
+                accelerators = {f'{k}:{v}' for k, v in accelerators.items()}
```
```suggestion
                accelerators = [f'{k}:{v}' for k, v in accelerators.items()]
```

MaoZiming replied:
updated

infwinston reviewed (commented):
Great PR!! left some more comments.

MaoZiming commented on examples/multi_resources.py:
```diff
@@ -0,0 +1,77 @@
+import sky
```
added tests to `skypilot/tests`

MaoZiming commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -2611,17 +2611,16 @@ def check_resources_fit_cluster(
                 does not match the existing cluster.
+
+        If multiple resources are specified, this will pass only when
+        there exists an exact match to the cluster.
```
Changed comment.
This will pass as long as there is one resource matching the cluster

MaoZiming commented:
@infwinston Thanks! I have included most of the comments. 
Currently, `sky exec` does not support multiple resources (can be next PR). 
Refactoring `resources` or `accelerators` also can be deferred to later.

Michaelvll reviewed (commented):
I haven't finished the pass of the whole PR yet, but I am wondering if the following way is possible to support a flexible multiple resources in our yaml file in additional to the list of accelerators and regions:

```yaml
# ordered
resources:
  preference: ordered
  candidates:
    - accelerators: V100
    - cloud: aws
      accelerators: A100
      cpus: 2
    - cloud: aws
```

```yaml
# nonordered
resources: 
  preference: any
  candidates:
    - accelerators: V100
    - cloud: aws
      accelerators: A100
      cpus: 2
    - cloud: aws
```

That said we still keep the original way to specify `resources` as a syntax sugar.
With this, we can modify the `from_config_dict` function of resources to create a list or a set of resources to match the python API?

Michaelvll commented on sky/execution.py:
```diff
@@ -567,6 +567,13 @@ def exec(  # pylint: disable=redefined-builtin
+    if isinstance(entrypoint, sky.Task):
+        resources = entrypoint.resources
+        if len(resources) > 1:
+            logger.warning(f'{colorama.Fore.YELLOW} Passing multiple resources '
+                           'to sky.exec is not supported.'
+                           f'{colorama.Style.RESET_ALL}')
+            return
```
Should we `raise` instead?

Also, this means we are not able to `sky exec` the same task yaml that is used for launching the cluster. Why can't we do the same checking we did in `cloud_vm_ray_backend::check_resources_fit_cluster`

Michaelvll commented on sky/optimizer.py:
```diff
@@ -56,6 +56,16 @@ def _create_table(field_names: List[str]) -> prettytable.PrettyTable:
     return log_utils.create_table(field_names, **table_kwargs)
 
 
+def _is_dag_with_res_list(dag: 'dag_lib.Dag') -> bool:
```
style nit: avoid abbreviation. reference:https://google.github.io/styleguide/pyguide.html#316-naming

MaoZiming commented:
@Michaelvll Hey, thanks for the reviews! I have addressed all the comments. I am thinking maybe we can use 
`accelerators: ['A100-40GB:1', 'K80:1', 'V100:1', 'T4:1']` for ordered list, and
`accelerators: {'A100-40GB': 1, 'K80': 1, 'V100': 1, 'T4': 1}` for unordered list.

```
accelerators: !!set
  ? A100-40GB:1
  ? V100:1
```
Should also work but it looks a bit awkward

Michaelvll commented:
> @Michaelvll Hey, thanks for the reviews! I have addressed all the comments. I am thinking maybe we can use `accelerators: ['A100-40GB:1', 'K80:1', 'V100:1', 'T4:1']` for ordered list, and `accelerators: {'A100-40GB': 1, 'K80': 1, 'V100': 1, 'T4': 1}` for unordered list.
> 
> ```
> accelerators: !!set
>   ? A100-40GB:1
>   ? V100:1
> ```
> 
> also work but it looks a bit weird

I agree the laste `!!set` way looks a bit weird. For the unordered list of accelerators, I think we can support both:
1. `{'A100-40GB': 1, 'K80': 1, 'V100': 1, 'T4': 1}`
2. `{'A100-40GB:1', 'K80:1', 'V100:1', 'T4:1'}`
What do you think?

MaoZiming commented:
@Michaelvll Agree
Pushed a new version.

Michaelvll commented on sky/optimizer.py:
```diff
@@ -114,11 +124,26 @@ def optimize(dag: 'dag_lib.Dag',
+                # Honor the user's choice.
+                # Take dag.tasks[0], the actual task
+                # dag can store dummy tasks.
+                resources_list = dag.tasks[0].resources
+                accelerators_str = ', '.join([
+                    r.get_accelerators_str() + r.get_spot_str()
+                    for r in resources_list
+                ])
+                logger.info(
+                    f'{colorama.Fore.YELLOW}Using user-specified accelerators list{colorama.Style.RESET_ALL} (will be tried in the listed order): {accelerators_str}'  # pylint: disable=line-too-long
```
It seems we can combine the handling of ordered resources and the normal optimization with the following algorithm (we may not need the _set_resources_by_user_order):

```
def _optimize_dag(dag, task_id):
    local_dag = copy.deepcopy(dag)
    task = dag.tasks[task_id]
    local_task = local_dag.tasks[task_id]
    if isinstance(task.resources, list):
        for resource in task.resources:
            local_task.set_resources(resource)
            if task_id == len(dag.tasks) - 1:
                plan = Optimizer._optimize_objective(local_dag,
                                            minimize_cost=minimize == OptimizeTarget.COST,
                                            blocked_resources=blocked_resources,
                                            quiet = quiet)
                if plan is not None:
                    return plan
            else:
                plan = _optimize_dag(local_dag, task_id + 1)
                if plan is not None:
                    return plan
plan = _optimize_dag(dag, 0)
_set_plan_to_best_resources(dag, plan)
```

MaoZiming replied:
I think we cannot combine that. Eventually, we want to print out an optimizer table showing different accelerators that the optimizer considered. 
if we do above, it will just show one. We still need to write more code

MaoZiming replied:
@Michaelvll I think we can leave with `assert len(dag.tasks) == 1, 'Multi-accelerators only Supports one task for the dag'`.
Optimizing multiple tasks for the dag previously require `_optimize_by_dp` or `_optimize_by_ilp` (by cost or time), which are not needed if we just follow user specified order.

MaoZiming replied:
Addressed. Multi-task dag will print every task's resources (if multiple resources are specified)
also refactored.

Michaelvll commented on sky/optimizer.py:
```diff
@@ -1026,7 +1212,8 @@ def _fill_in_launchable_resources(
                     launchable[resources].extend(
-                        _make_launchables_for_valid_region_zones(cheapest))
+                        _make_launchables_for_valid_region_zones(
+                            cheapest, resources.region_list))
```
Instead of adding a new API `region_list` in the `Resources`, can we do a similar thing as list of accelerators, i.e., creating multiple `Resources` object with a single region as candidate in the upper level callers?

MaoZiming replied:
I think the tricky thing is when user specifies a list for both regions and accelerators. Do we do a cartesian product?

MaoZiming replied:
same question as next one. commented out.

Michaelvll commented on sky/resources.py:
```diff
@@ -56,7 +56,7 @@ def __init__(
         use_spot: Optional[bool] = None,
         spot_recovery: Optional[str] = None,
-        region: Optional[str] = None,
+        region: Optional[Union[str, List[str]]] = None,
```
As mentioned above, we may not want to add an additional API for `Resources`. Why not do similar thing as `accelerators`, i.e., when user specifies multiple regions, we create multiple `Resources` object that has a single region specified?

MaoZiming replied:
The reason is because when the user specifies a list of accelerators, and a list of regions, we are not sure if we want to do a cartesian product between the two

MaoZiming replied:
removed regions from this PR

Michaelvll commented on sky/spot/controller.py:
```diff
@@ -48,7 +48,6 @@ def __init__(self, job_id: int, dag_yaml: str,
                  retry_until_up: bool) -> None:
         self._job_id = job_id
         self._dag, self._dag_name = _get_dag_and_name(dag_yaml)
-        logger.info(self._dag)
```
nit: should we keep this?

Michaelvll commented on tests/test_multi_resources.py:
```diff
@@ -0,0 +1,39 @@
+import sky
```
Should we have it in the optimizer test instead? Also, do we want to add it to CI?

MaoZiming replied:
moved.

Michaelvll commented on sky/utils/schemas.py:
```diff
@@ -50,8 +57,16 @@ def get_resources_schema():
+                    }
                 }]
             },
+            'resources_ordered': {
+                'type': 'boolean',
+            },
```
Need to update?

Michaelvll reviewed (commented):
Thanks for the great effort and the update @MaoZiming! I went through rest of the files, except for the spot related ones and left some comments for the design and cleanness of the code.  : )

cblmemo commented on sky/execution.py:
```diff
@@ -573,6 +573,11 @@ def exec(  # pylint: disable=redefined-builtin
         dryrun=dryrun)
+
+    if isinstance(entrypoint, sky.Task):
+        backend.check_resources_fit_cluster(handle, task, check_ports=True)
```
Humm, seems like the comment is outdated. @MaoZiming did you move this part of the code to other places, or did you remove it?

MaoZiming replied:
yeah I removed it last week. another place is 
https://github.com/skypilot-org/skypilot/pull/2498/files#diff-f14940b9d929546868fb41cb0b27a1db08b321b7bbc0cb3ab3b864589aa82aa8L3484

cblmemo replied:
So it is the same with the master branch?

MaoZiming replied:
yep

MaoZiming commented on sky/backends/backend_utils.py:
```diff
@@ -1947,46 +1947,51 @@ def check_can_clone_disk_and_override_task(
-    else:
-        if not original_cloud.is_same_cloud(task_resources.cloud):
+    new_task_resources = []
+    for task_resources in task.resources:
```
Let me put the print exception to the end.

MaoZiming commented on sky/backends/backend_utils.py:
```diff
@@ -2690,17 +2695,19 @@ def get_backend_from_handle(
-    return accelerator_dict
+    else:
+        # Multiple resources specified.
+        accelerator_dict = {}
+        for resource in task.resources:
+            accelerator_dict.update(resource.accelerators)
```
@Michaelvll I think so. Do we want to both allow `V100:1` and `V100:4`? In that case, using dict is probably not good. Is there a use case where user wants to use either `V100:1` or `V100:4`?

MaoZiming replied:
fixed

MaoZiming commented on sky/cli.py:
```diff
@@ -1128,23 +1128,20 @@ def _make_task_or_dag_from_entrypoint_with_overrides(
-        if acc.startswith('tpu-') and task.num_nodes > 1:
-            raise ValueError('Multi-node TPU cluster is not supported. '
-                             f'Got num_nodes={task.num_nodes}.')
+    for res in task.resources:
+        if res.accelerators is not None:
+            acc, _ = list(res.accelerators.items())[0]
+            if acc.startswith('tpu-') and task.num_nodes > 1:
+                raise ValueError('Multi-node TPU cluster is not supported. '
+                                 f'Got num_nodes={task.num_nodes}.')
```
moved

MaoZiming commented on sky/optimizer.py:
```diff
@@ -775,19 +864,28 @@ def format_number(x):
-                                                           v[chosen_resources])
-
+            chosen_cost = 0.0
+            for resources, cost in v.items():
+                if resources.to_yaml_config(
+                ) == chosen_resources.to_yaml_config():
+                    chosen_cost = cost
+                    break
```
The new changes are making `resources` not hashable. I can probably do a custom hash function

MaoZiming replied:
It seems the resources.cloud field are different. The difference is in `TTLCache`
```
_list_reservations_cache: TTLCache({('n1-highmem-4',): []}, maxsize=1, currsize=1)
_list_reservations_cache: None
```

MaoZiming replied:
fixed

MaoZiming commented on sky/optimizer.py:
```diff
@@ -804,7 +902,16 @@ def format_number(x):
+            if isinstance(task.resources, list):
+                rows = sorted(
+                    rows,
+                    key=lambda x: (
+                        resources_list.index(x[-4] + (  # pylint: disable=cell-var-from-loop
+                            '[Spot]' if '[Spot]' in x[1] else '')),  # pylint: disable=cell-var-from-loop
+                        x[-2]))  # pylint: disable=cell-var-from-loop
```
done

MaoZiming commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -3482,6 +3497,8 @@ def _execute(
         # We need to check ports here since sky.exec shouldn't change resources
         self.check_resources_fit_cluster(handle, task, check_ports=True)
+        if handle.launched_resources:
+            task.set_resources(handle.launched_resources)
```
@Michaelvll I am trying to handle the case where we exec on an existing cluster with a multi-resource yaml, I want to overwrite the multiple resources with the resource of the existing cluster. I thought the previous cluster cannot be in the INIT state by the time we reach here?

MaoZiming replied:
fixed

Michaelvll commented on sky/task.py:
```diff
@@ -422,10 +430,39 @@ def from_yaml_config(
-        task.set_resources({resources})
+        resources_config = config.pop('resources', None)
+        # Translate accelerators field to potential multiple resources.
+        if resources_config and resources_config.get(
+                'accelerators') is not None:
```
If we don't have a way to take a set of resources not only having the accelerators different, e.g. `{Resources(sky.GCP()), Resources(sky.AWS())}`, our python API for spot instances will fail. As our `spot_launch` relies on the dumping and reloading for the yaml file. Should we support the list of resources support in yaml as well?
(We may not need to expose this to the user at the moment if we are not sure with the interface yet, but this should enable us to dump and reload any resources specification in python API)
```
resources:
  any_of:
    - cloud: AWS
       accelerator: A10g
    - cloud: GCP
       accelerator: L4
```
and for unordered list:
```
resources:
  ordered:
    - cloud: AWS
       accelerator: A10g
    - cloud: GCP
       accelerator: L4
```

MaoZiming replied:
Done. Please see `examples/multi_resources.yaml`

Michaelvll commented on sky/task.py:
```diff
@@ -592,8 +629,18 @@ def set_resources(
+            new_resources = res.copy(**override_params)
+            new_resources_list.append(new_resources)
+
+        if isinstance(self.resources, list):
+            self.set_resources(new_resources_list)
+        else:
+            self.set_resources(set(new_resources_list))
```
nit: `type(self.resources)(new_resources_list)`

Michaelvll reviewed (commented):
Thanks for the update @MaoZiming! This is excellent. Left several comments for the optimizer and the yaml representation. We probably want to support multiple resources in yaml as well.

MaoZiming commented on sky/optimizer.py:
```diff
@@ -887,6 +985,58 @@ def _optimize_objective(
         return best_plan
 
+    @staticmethod
+    def _optimize_by_user(
```
Thanks! I have pushed the changes for [#2498 (comment)](https://github.com/skypilot-org/skypilot/pull/2498#discussion_r1363368294)

MaoZiming commented on sky/resources.py:
```diff
@@ -1252,3 +1264,11 @@ def __setstate__(self, state):
+        return str(vars(self)) == str(vars(other))
+
+    def __hash__(self):
+        return hash(str(vars(self)))
```
I couldn't really think of a case where this can cause conflict. @Michaelvll

MaoZiming commented:
@Michaelvll Thanks for the code review! I have resolved all the suggestions.

MaoZiming commented:
@Michaelvll Pushed a fix. The code should be ready for review.

Michaelvll commented on sky/optimizer.py:
```diff
@@ -705,12 +798,7 @@ def print_optimized_plan(
-                accelerators, count = list(accelerators.items())[0]
-                accelerators = f'{accelerators}:{count}'
+            accelerators = resources.get_accelerators_str()
             spot = '[Spot]' if resources.use_spot else ''
```
Should we use `resources.get_spot_str()`?

Michaelvll commented on sky/optimizer.py:
```diff
@@ -740,6 +828,55 @@ def format_number(x):
+                                       cost_str: str, chosen: bool) -> Row:
+
+            accelerators = resources.get_accelerators_str()
+            spot = '[Spot]' if resources.use_spot else ''
```
Should we use `resources.get_spot_str()`

Michaelvll commented on sky/spot/recovery_strategy.py:
```diff
@@ -94,16 +94,17 @@ def __init_subclass__(cls, name: str, default: bool = False):
-            'spot_recovery is required to use spot strategy.')
+        resource_list = list(task.resources)
+        spot_recovery = resource_list[0].spot_recovery
+        if spot_recovery is None:
+            spot_recovery = SPOT_DEFAULT_STRATEGY
```
`spot_recovery` should be guaranteed to be not None in `spot_launch`?

Michaelvll replied:
Also, should we assert that all the resources have the same `spot_recovery`?

Michaelvll commented on sky/spot/recovery_strategy.py:
```diff
@@ -394,8 +396,10 @@ def _launch(self,
             launched_resources = handle.launched_resources
             self._launched_cloud_region = (launched_resources.cloud,
                                            launched_resources.region)
+            self._launched_resources = launched_resources  # pylint: disable=line-too-long
```
```suggestion
            self._launched_resources = launched_resources
```

Michaelvll commented on sky/spot/recovery_strategy.py:
```diff
@@ -394,8 +396,10 @@ def _launch(self,
             launched_resources = handle.launched_resources
             self._launched_cloud_region = (launched_resources.cloud,
                                            launched_resources.region)
```
Do we still need this if `self._launched_resources` is set?

Michaelvll commented on sky/spot/recovery_strategy.py:
```diff
@@ -412,18 +416,17 @@ def recover(self) -> float:
                 launched_cloud, launched_region = self._launched_cloud_region
-                new_resources = resources.copy(cloud=launched_cloud,
-                                               region=launched_region)
+                new_resources = self._launched_resources.copy(
+                    cloud=launched_cloud, region=launched_region)
```
It seems we don't need to set the cloud and region again here, but instead, we may want to set the `zone` to None?

Michaelvll commented on sky/task.py:
```diff
@@ -422,10 +427,49 @@ def from_yaml_config(
-        task.set_resources({resources})
+        resources_config = config.pop('resources', None)
+        if resources_config and resources_config.get('any_of') is not None:
+            resources_set = set()
```
We should `raise ValueError` (or use the schema check) when there are additional arguments provided in the `resources` field, e.g.,
```
if len(resources_config) > 1:
  raise ValueError(...)
```

Michaelvll commented on sky/task.py:
```diff
@@ -961,10 +1015,21 @@ def add_if_not_none(key, value, no_empty: bool = False):
-            resources = list(self.resources)[0]
-            add_if_not_none('resources', resources.to_yaml_config())
+        tmp_resource_config = {}
+        if isinstance(self.resources, list):
+            resource_list = []
+            for r in self.resources:
+                resource_list.append(r.to_yaml_config())
+            tmp_resource_config['ordered'] = resource_list
+        elif len(self.resources) > 1:
+            resource_list = []
+            for r in self.resources:
+                resource_list.append(r.to_yaml_config())
+            tmp_resource_config['any_of'] = resource_list
```
```suggestion
        if isinstance(self.resources, (list, set)):
            resource_list = []
            for r in self.resources:
                resource_list.append(r.to_yaml_config())
            key = 'ordered' if isinstance(self.resources, list) else 'any_of'
            tmp_resource_config[key] = resource_list
```

Michaelvll commented on sky/utils/dag_utils.py:
```diff
@@ -94,18 +94,31 @@ def maybe_infer_and_fill_dag_and_task_names(dag: dag_lib.Dag) -> None:
+                    'being launched. Ignoring the field and proceeding to use spot '  # pylint: disable=line-too-long
+                    'instance(s).')
+            change_default_value['use_spot'] = True
+            if resources.spot_recovery is None:
+                change_default_value[
+                    'spot_recovery'] = spot.SPOT_DEFAULT_STRATEGY
```
We probably want to make sure the spot recovery config for all the resources are the same.

Michaelvll commented on sky/utils/schemas.py:
```diff
@@ -50,6 +50,11 @@ def get_resources_schema():
                     'additionalProperties': {
                         'type': 'number'
                     }
+                }, {
+                    'type': 'array',
+                    'items': {
+                        'type': 'string',
+                    }
```
Will there be a difference between `[V100:1, A100:1]` vs `{V100:1, A100:1}` here?

MaoZiming replied:
@Michaelvll  'array` only applies to `['A100-40GB:1', 'V100:1', 'K80:1', 'T4:1']`
`{'A100-40GB:1', 'K80:1', 'V100:1', 'T4:1', 'T4:4'}` will be read as a dictionary.

Michaelvll replied:
I see, in that case, which of the schemas here represents the `{V100:1, A100:1}`?

MaoZiming replied:
`{V100:1, A100:`}` is not standard json, so it got checked with `type: string`

Michaelvll replied:
Oh, that is a bit suprising as our code will read this as a dict as you mentioned. It worth a comment here.

MaoZiming replied:
@Michaelvll let me add a comment there. thanks!

Michaelvll reviewed (commented):
Thanks for the quick fix @MaoZiming! The code looks mostly good to me now. Left several comments. Once they are fixed, I am going to do some test on it. ; )

MaoZiming commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -2725,40 +2727,54 @@ def check_resources_fit_cluster(
+                            f'{example_resource.zone!r},'
+                            'but the existing cluster '
+                            f'{zone_str}')
                 with ux_utils.print_exception_no_traceback():
                     raise exceptions.ResourcesMismatchError(
```
@Michaelvll If we resources can have different zones, we need to add that. (I know we don't support multiple zones yet)

MaoZiming replied:
moved out of the for loop

MaoZiming commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -2725,40 +2727,54 @@ def check_resources_fit_cluster(
+                break
+            else:
+                requested_resource_list.append(f'{task.num_nodes}x {resource}')
+        requested_resource_str = ', '.join(requested_resource_list)
+        if isinstance(task.resources, list):
+            requested_resource_str = f'[{requested_resource_str}]'
+        elif isinstance(task.resources, set):
+            requested_resource_str = f'{{{requested_resource_str}}}'
```
@Michaelvll It seems we need to `', '` join each resource first then add `[]` or `{}`?

MaoZiming commented on sky/optimizer.py:
```diff
@@ -114,11 +184,29 @@ def optimize(dag: 'dag_lib.Dag',
+                        )
+                _ = Optimizer._optimize_objective(
+                    dag=dag,
+                    optimize_target=OptimizeTarget.USER,
```
fixed

MaoZiming commented on sky/optimizer.py:
```diff
@@ -851,34 +1005,45 @@ def _optimize_objective(
+                dag=dag,
+                task_id=0,
+                optimize_target=OptimizeTarget.COST,
+                blocked_resources=blocked_resources)
```
fixed

MaoZiming commented:
@Michaelvll Addressed the comments. Thanks for the suggestions!

Michaelvll commented on sky/utils/dag_utils.py:
```diff
@@ -94,18 +94,30 @@ def maybe_infer_and_fill_dag_and_task_names(dag: dag_lib.Dag) -> None:
+        spot_recovery_strategy = new_resources_list[0].spot_recovery
+        for resource in new_resources_list:
+            if resource.spot_recovery != spot_recovery_strategy:
+                raise ValueError('All resources in the task must have'
```
```suggestion
                with ux_utils.print_no_traceback():
                    raise ValueError('All resources in the task must have'
```

Michaelvll commented on sky/utils/schemas.py:
```diff
@@ -106,6 +111,18 @@ def get_resources_schema():
                 }]
+            },
+            'any_of': {
+                'type': 'array',
+                'items': {
+                    'type': 'string',
+                },
```
We are taking a list of `resources` schema. How about we define the schema for a single `resources` and have it here?

MaoZiming replied:
added

Michaelvll reviewed (commented):
Thanks for the update @MaoZiming! The code looks mostly good to me! Once we have done the test for the smoke test and some manual tests. I think it should be good to go in.

MaoZiming commented on sky/optimizer.py:
```diff
@@ -1008,42 +946,94 @@ def _optimize_objective(
         # TODO: The output of this function is useful. Should generate a
         # text plan and print to both console and a log file.
 
+        is_resources_ordered = _is_dag_resources_ordered(dag)
```
sure!

MaoZiming commented on sky/task.py:
```diff
@@ -428,12 +428,26 @@ def from_yaml_config(
+                    'Cannot specify both "all_of" and "ordered" in resources.')
         if resources_config and resources_config.get('any_of') is not None:
+            if len(resources_config) > 1:
+                with ux_utils.print_exception_no_traceback():
```
Added a TODO comment here.

Michaelvll reviewed (commented):
Thanks for the update @MaoZiming! I tried the following, but an error occurred:
```yaml
resources:
  any_of:
    - cloud: aws
      region: us-east-1
      accelerators: A100:8
    - cloud: gcp
      accelerators: T4:4
    - cloud: aws

run:
  echo hi
```
`sky launch -c test test.yaml` twice:

```
sky/backends/cloud_vm_ray_backend.py", line 4347, in _check_existing_cluster
    assert len(task.resources) == 1
AssertionError
```

MaoZiming commented on sky/optimizer.py:
```diff
@@ -946,8 +949,8 @@ def _optimize_dag(
 
-        is_resources_ordered = _is_dag_resources_ordered(dag)
-        if is_resources_ordered:
+        has_resources_ordered = _is_dag_resources_ordered(dag)
```
this should already be fixed

MaoZiming commented on sky/optimizer.py:
```diff
@@ -851,40 +946,106 @@ def _optimize_objective(
+                    resources_list = task.resources
+                    accelerators_str = ', '.join(
+                        [f'{r}' for r in resources_list])
+                    logger.info(f'{colorama.Fore.YELLOW}{task_id + 1}-th '
```
sure!

MaoZiming commented on sky/utils/schemas.py:
```diff
@@ -114,13 +219,13 @@ def get_resources_schema():
                 'type': 'array',
                 'items': {
-                    'type': 'string',
+                    'type': 'object',
```
let me add this.

MaoZiming commented:
@Michaelvll Thanks! pushed a fix

Michaelvll commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -3508,12 +3508,13 @@ def _execute(
         task_copy = copy.copy(task)
         # Handle multiple resources exec case.
         task_copy.set_resources(valid_resource)
+        logger.info(valid_resource)
```
Should we elaborate this output? It would be a bit surprising to see just a resources in the output.
For example:
```suggestion
        if len(task.resources) > 1:
            logger.info(f'Multiple resources are specified for the task, using: {valid_resource}')
```

Michaelvll commented on sky/optimizer.py:
```diff
@@ -851,40 +946,124 @@ def _optimize_objective(
+                    accelerators_str = ', '.join(
+                        [f'{r}' for r in resources_list])
+                    task_id_str = ordinal_number(task_id + 1)
+                    # Add in dummy tasks
+                    if len(dag.tasks) > 3:
```
```suggestion
                    if len(dag.tasks) > 3:
                    # User-provided dag has more than one task. Comparing with 3,
                    # as there are two dummy tasks added by the optimizer.
                    # by the optimizer.
```

Michaelvll commented on tests/test_multi_regions.py:
```diff
@@ -0,0 +1,37 @@
+import sky
```
This seems not relavant any more?

Michaelvll commented on tests/test_smoke.py:
```diff
@@ -3558,3 +3558,32 @@ def test_load_dump_yaml_config_equivalent(self):
+
+
+# ---------- Testing Multiple Resources ----------
+@pytest.mark.gcp
```
We probably do not need this?

Michaelvll commented on tests/test_smoke.py:
```diff
@@ -3558,3 +3558,32 @@ def test_load_dump_yaml_config_equivalent(self):
+    run_one_test(test)
+
+
+@pytest.mark.gcp
```
Same, we don't need this?

Michaelvll commented on tests/test_smoke.py:
```diff
@@ -3558,3 +3558,32 @@ def test_load_dump_yaml_config_equivalent(self):
+        ],
+        f'sky down -y {name}',
+    )
+    run_one_test(test)
```
Do we want to add another test for the multiple resources?

MaoZiming replied:
Sure!

Michaelvll reviewed (approved):
Thanks for the great effort @MaoZiming! The LGTM now. Once it has passed the smoke test and any additional manual tests, it should be good to go in. : )

MaoZiming commented:
Thanks for the review! Addressed the comments. 
The smoke tests should be good (except the ones I mentioned in the issue)
